### ISSUE TITLE
Priority of package-stubs PYI files over package PYI files

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#3540]
**Describe the bug**
I'm working on creating the new pandas-stubs repo.  The goal is to provide a wheel that will install as the package `pandas-stubs`. There are PYI files in there that are different than some of the PYI files that are shipped with pandas.  If I install my working copy of `pandas-stubs` into my environment, then if there is a PYI file in the pandas distribution, that is taking precedence over the PYI file in the `pandas-stubs` package.  It should be the other way around, where the PYI files in `pandas-stubs` take precedence.

**To Reproduce**

1. Pull https://github.com/pandas-dev/pandas-stubs as of commit https://github.com/pandas-dev/pandas-stubs/commit/5f8cdaa78c5a654e7156ab9ee2473e91627a4dcd
2. Create an environment based on `tests/requirements.txt` in that repo.  
3. Build the wheel `python setup.py build bdist_wheel`
4. Install the wheel `pip install dist/pandas_stubs-1.4.2.220604-py3-none-any.whl`
5. Run `pyright -p pyrightdistconfig.json`

What happens is that `pyright` is finding the PYI files in `pandas/_libs/tslibs/*.pyi` instead of the ones in `pandas-stubs/_libs/tslibs/*.pyi`, which will eventually be the public PYI files.  The ones shipped with pandas are used for internal testing of pandas and should be ignored by users.  If you remove the files in `pandas/_libs/tslibs/*.pyi`, then the tests pass fine.

**Expected behavior**
Any stub present in `pandas-stubs` should override a stub present in the installed version of `pandas`.

**VS Code extension or command-line**

pyright version 1.1.252

**Additional context**
Things work fine with `mypy` in terms of it using the stubs in `pandas-stubs` when they are installed.

One other note.  It would be useful if when using the `--dependencies` flag, `pyright` printed which stub it was using.  I get output like this in the CI:  `/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/pandas/util/_print_versions.pyi` .  That PYI file is not in `pandas` - it is in `pandas-stubs`.  pyright should indicate that it is using the file in `pandas-stubs` .

Also, the failure in CI when using pyright is shown here:
https://github.com/pandas-dev/pandas-stubs/runs/6743066046?check_suite_focus=true

When I added to the CI a hack to remove `pandas/_libs/tslibs/*.pyi`, then the successful run is shown here:  https://github.com/pandas-dev/pandas-stubs/runs/6743079731?check_suite_focus=true

**Comment by erictraut** [comment]
What is the symptom you are seeing? I've followed the steps provided, and running pyright produces no type errors.

```
Loading configuration file at /Users/eric/Desktop/pandas-stubs/pyrightdistconfig.json
Assuming Python version 3.10
Assuming Python platform Darwin
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
Searching for source files
Found 10 source files
0 errors, 0 warnings, 0 informations 
Completed in 2.287sec
```

I used Python 3.10 and created a venv from scratch (located outside of the workspace). I then activated the venv and followed the steps listed above.

I confirmed that I also received no type errors when viewing the test source files in VS Code with the pyright language server installed.

I noticed that you have a "typings/pandas" directory checked in to this repo. Was this intended? A local "typings" directory will always take precedence over stub libraries or normal libraries installed in the selected python environment. The import resolution order used by pyright is documented [here](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md) and conforms to [PEP 561](https://peps.python.org/pep-0561/).

**Comment by Dr-Irv** [comment]
Ah, it was 2AM in the morning, and I left out one key step.

After you build the wheel, remove the `typings` directory.  Alternatively, do something like this:
```bash
cd pandas-stubs
mkdir ../timetest
mv tests/pandas/test_timefuncs.py ../timetest
cd ../timetest
pyright test_timefuncs.py
```

The idea here is to get pyright to not look at the `typings` folder, and have it look for the types in the installed `pandas-stubs` package.

The first error you will get is this:
```
test_timefuncs.py:16:24 - error: Expression of type "Timestamp | NaTType" cannot be assigned to declared type "Timestamp"
    Type "Timestamp | NaTType" cannot be assigned to type "Timestamp"
      "NaTType" is incompatible with "Timestamp" (reportGeneralTypeIssues)
```
That's because it is reading the pandas version of `pandas/_libs/tslibs/timestamps.pyi` instead of `pandas-stubs/_libs/tslibs/timestamps.pyi` .  Right now, `pandas/_libs/tstlibs/timestamps.pyi` has a `__new__()` that returns `_DatetimeT | NaTType` while `pandas-stubs/_libs/tslibs/timestamps.pyi` has `__new__()` that returns `_DatetimeT` .

**Comment by Dr-Irv** [comment]
> The import resolution order used by pyright is documented [here](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md) and conforms to [PEP 561](https://peps.python.org/pep-0561/).

That says:

- For a given package, try to resolve first using a stub package. Stub packages, as defined in [PEP 561](https://www.python.org/dev/peps/pep-0561/#type-checker-module-resolution-order), are named the same as the original package but with “-stubs” appended.
- Try to resolve using an inline stub, a “.pyi” file that ships within the package.

What I think is happening is that if there is a `.pyi` file in the package, that is taking precedence over the `.pyi` file in the `-stubs` package, and it should be the other way around.

**Comment by erictraut** [comment]
@jakebailey, this appears to be a bug in the [virtual file system](https://github.com/microsoft/pyright/blob/ac576bbd479bcd51dbfadfc6147f5da6806217c3/packages/pyright-internal/src/pyrightFileSystem.ts#L6) that you added for pylance last year. Can you remind me again why we are treating partial stub packages "as though they are copied into the associated package, and the combined set of files is treated as one"? It looks like the logic assumes that there will be no overlaps between the partial stub package and the non-stub package. If there are overlaps (i.e. there is a `.pyi` file with the same name in both locations), it resolves to the wrong one. As @Dr-Irv says above, the `.pyi` in the partial stub package should take precedent.

Also, this virtual file system cleverness is resulting in confusing output for the "--dependencies" command-line option because the resolved imports are being reported with their virtual file system paths rather than the real file system paths. In other words, an implementation detail is leaking.

**Comment by erictraut** [comment]
Oh, I think I just remembered why you added this virtual file system "overlay". If I remember correctly, it was to handle relative imports within the package. Is that right?

**Comment by jakebailey** [comment]
Relative accesses are a part of this, yes.

> Can you remind me again why we are treating partial stub packages "as though they are copied into the associated package, and the combined set of files is treated as one"?

The PEP says:

> Type checkers should merge the stub package and runtime package or typeshed directories. This can be thought of as the functional equivalent of copying the stub package into the same directory as the corresponding runtime package or typeshed folder and type checking the combined directory structure.

The difference being that we don't physically do this, but instead use the VFS to simulate this operation so that the rest of the code doesn't know this is happening. Relative imports are a part of that, because they need to still resolve properly, but I seem to remember there being bugs in the past around which import you got when you were in a partial stub and did an absolute vs relative import (@heejaechang would remember).

But, it sounds like what you're seeing is that we sometimes _don't_ overlay the stubs on top of the real package, and end up getting the real package's stubs?

Is pandas marked as `py.typed`, or is this in the situation where `pandas` is still marked as untyped?  If so, that's when you start getting into the complicated mess of priorities and deciding whether or not to use a particular package's own types, etc, which we do differently than other checkers AFAIK.

**Comment by erictraut** [comment]
I've fixed the `--dependencies` command so it outputs the original file system path rather than the internal virtual file system path. That bug resulted in confusing output, but it doesn't explain the underlying problem you're seeing.

With the `--dependencies` bug fixed, everything now appears correct to me. All of the dependencies refer to stubs in the "site-packages/pandas-stubs" directory rather than the "site-packages/pandas" directory, and I receive no type checking errors. @Dr-Irv, can you confirm that you are seeing actual type checking errors? It looks from your initial bug report that you are, but I'm not able to repro that. Any other suggestions?

@jakebailey asked:
> Is pandas marked as py.typed?

In this case, pandas is not marked py.typed, but pandas-stubs contains a py.typed file with "partial".

**Comment by Dr-Irv** [comment]
> With the --dependencies bug fixed, everything now appears correct to me. All of the dependencies refer to stubs in the "site-packages/pandas-stubs" directory rather than the "site-packages/pandas" directory, and I receive no type checking errors. @Dr-Irv, can you confirm that you are seeing actual type checking errors? It looks from your initial bug report that you are, but I'm not able to repro that. Any other suggestions?

1) Did you do the repro step above where you either remove the `typings` folder or move the tests to a different directory?
2) Do I need the new version to test?
3) It may be the `py.typed` file creating the issue so I can remove that

**Comment by Dr-Irv** [comment]
> It may be the py.typed file creating the issue so I can remove that

That was the issue.  I'll keep that out of the distribution.  It was there in the other `pandas-stubs` project, but I think that's because when you install that project, it installs directly into the `pandas` installation.

**Comment by Dr-Irv** [comment]
> > It may be the py.typed file creating the issue so I can remove that
> 
> That was the issue. I'll keep that out of the distribution. It was there in the other `pandas-stubs` project, but I think that's because when you install that project, it installs directly into the `pandas` installation.

Well, removing `py.typed` messes up the `mypy` tests.  So ideally having `py.typed` in `pandas-stubs` should only apply to the stubs, not the installed package.

**Comment by erictraut** [comment]
To be clear, you don't have a "py.typed" in the installed "pandas" package, right? There's a "py.typed" only within the "pandas-stubs" package, and it contains "partial", indicating that the type stubs are partial, right? That should work fine.

To answer your questions above... yes, I followed the repro steps and removed the `typings` folder. I also tried moving the tests to a different directory. The result was the same in both cases — no errors when I ran pyright.

You'll need a new version of pyright for the "--dependencies" bug fix, but that was just a cosmetic bug and didn't affect the type checking logic.

If you're still seeing errors when you run pyright, then there's still an underlying issue that we need to understand. I'm not able to repro the problem as you've described it. Could you start with a clean virtual environment and see if you can repro it there?

**Comment by Dr-Irv** [comment]
> To be clear, you don't have a "py.typed" in the installed "pandas" package, right? There's a "py.typed" only within the "pandas-stubs" package, and it contains "partial", indicating that the type stubs are partial, right? That should work fine.
> 

That's correct.


> To answer your questions above... yes, I followed the repro steps and removed the `typings` folder. I also tried moving the tests to a different directory. The result was the same in both cases — no errors when I ran pyright.

So differences could be that I'm on Windows, but I've also got it reproduced in the CI.  See below.
> 

> 
> If you're still seeing errors when you run pyright, then there's still an underlying issue that we need to understand. I'm not able to repro the problem as you've described it. Could you start with a clean virtual environment and see if you can repro it there?

I use conda, so I created a conda environment using the following `env.yml` to run pyright in.
```
name: pstubtst
channels:
  - conda-forge
dependencies:
  - python=3.9
  - pandas=1.4.2
  - pyright
  - pytest
  - mypy=0.960
  - typing_extensions=4.2.0
  - matplotlib>=3.3.2
  - numpy>=1.22.3
  - setuptools>=62.3.2
  - wheel>=0.37.0
```
Copied the test files to a new directory, activated that environment, and got the same failure.  pyright is version 1.1.252.  Using python 3.9 on Windows.

Also, enabled the test in the CI, which is on Linux.  You should be able to get this branch from my repo: https://github.com/Dr-Irv/pandas-stubs/tree/pyrighttst

Then the result of the test is shown here:
https://github.com/Dr-Irv/pandas-stubs/runs/6748681574?check_suite_focus=true

What you will see (as compared to the `main` branch) is that the only difference is adding a pyright test on the installed wheel:
```
     - name: Build wheel and install and remove typings
        run: |
          python setup.py build bdist_wheel
          find ./dist/*.whl | xargs pip install
          rm -rf typings
          pip install pyright
      
      - name: run pyright command line
        run: |
          pyright --version
          pyright --dependencies tests/pandas/test_timefuncs.py
```

The first part is in `main`.  The second part is just running `pyright` on the one file, and you can see from the nature of the errors that it looked at the `pyi` file in `pandas/_libs/tslibs/timestamp.pyi` rather than `pandas-stubs/_libs/tslibs/timestamp.pyi` .  The wheel includes `py.typed` in `pandas-stubs` but not `pandas`.

Venturing a guess that you are using a Mac, so maybe that accounts for the difference?

**Comment by erictraut** [comment]
Yes, I'm on a Mac. I suppose it's possible that this is a platform-specific behavior, but it's unlikely.

For the record, here's the script that I'm running to repro:
```
# Activate Python 3.9
source ~/venvs/p39/bin/activate

# Create a fresh virtual environment
mkdir pandas-issue
cd pandas-issue
python -m venv venv
source venv/bin/activate

# Clone the repo
git clone --branch pyrighttst https://github.com/Dr-Irv/pandas-stubs
cd pandas-stubs

# Install requirements
pip install -r tests/requirements.txt 

# Build the wheel
python setup.py build bdist_wheel

# Install the wheel
pip install dist/pandas_stubs-1.4.2.220604-py3-none-any.whl

# Delete local stubs
rm -rf ./typings

# Run type checker
pyright -p pyrightdistconfig.json
```

When I run this, pyright emits no errors.

Interestingly, if I create the venv within the project directory, the `pandas-stubs` apparently doesn't get installed correctly. It doesn't appear in the `site-packages` directory. So perhaps there's a bug in the setup for the package?

Here's a slightly modified script that repros this:

```python
# Activate Python 3.9
source ~/venvs/p39/bin/activate

mkdir pandas-issue
cd pandas-issue

# Clone the repo
git clone --branch pyrighttst https://github.com/Dr-Irv/pandas-stubs
cd pandas-stubs

# Create a fresh virtual environment
python -m venv venv
source venv/bin/activate

# Install requirements
pip install -r tests/requirements.txt 

# Build the wheel
python setup.py build bdist_wheel

# Install the wheel
pip install dist/pandas_stubs-1.4.2.220604-py3-none-any.whl

# Delete local stubs
rm -rf ./typings

# Run type checker
pyright -p pyrightdistconfig.json
```

Here are the resulting errors in this case:
```
Found 10 source files
pandas-stubs/tests/pandas/test_indexes.py
  pandas-stubs/tests/pandas/test_indexes.py:20:17 - error: "assert_type" mismatch: expected "NDArray[bool_]" but received "Unknown"
pandas-stubs/tests/pandas/test_interval.py
  pandas-stubs/tests/pandas/test_interval.py:55:17 - error: "assert_type" mismatch: expected "int" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:56:17 - error: "assert_type" mismatch: expected "int" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:57:17 - error: "assert_type" mismatch: expected "int" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:58:17 - error: "assert_type" mismatch: expected "float" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:72:17 - error: "assert_type" mismatch: expected "float" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:73:17 - error: "assert_type" mismatch: expected "float" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:74:17 - error: "assert_type" mismatch: expected "float" but received "Unknown"
  pandas-stubs/tests/pandas/test_interval.py:75:17 - error: "assert_type" mismatch: expected "float" but received "Unknown"
9 errors, 0 warnings, 0 informations 
```

@debonte, you use a Windows system for your development, right? If so, could you see if you could try to repro the problem @Dr-Irv is seeing?

**Comment by Dr-Irv** [comment]
A few comments:
1) I did your first script on Windows and it failed.  However, the pip virtual environment is using a python from anaconda, so I don't know if that affects things.
2) For the test, you can just do `pyright tests/pandas/test_timefuncs.py` and if you see this, you know it is wrong:
```
*REDACTED*\pandas-issue\pandas-stubs\tests\pandas\test_timefuncs.py:16:24 - error: Expression of type "Timestamp | NaTType" cannot be assigned to declared type "Timestamp"
    Type "Timestamp | NaTType" cannot be assigned to type "Timestamp"
      "NaTType" is incompatible with "Timestamp" (reportGeneralTypeIssues)
```
3) `pandas-stubs` was installed in the right place:
```
 Directory of *REDACTED*\pandas-issue\venv\lib\site-packages

06/06/2022  10:34 AM    <DIR>          pandas
06/06/2022  10:34 AM    <DIR>          pandas-1.4.2.dist-info
06/06/2022  10:35 AM    <DIR>          pandas-stubs
06/06/2022  10:35 AM    <DIR>          pandas_stubs-1.4.2.220604.dist-info
```

**Comment by erictraut** [comment]
> It now is just finding the stubs in the local directory

Which stubs are you talking about? You've removed the `typings` contents, and the `pandas-stubs` are not installed, so I don't see how it could be finding stubs in the local directory. It's probably finding stubs that you have installed in your `site-packages`. You need to be careful when you run tests like this to ensure that you're using a virtual environment with a known state. That's why my script above sets up a new venv and installs everything from scratch.

You mentioned that you have conducted tests locally on Windows and in your Linux-based CI environment. Are you seeing the same results? If so, that tells us it's not related to the platform or case sensitivity of the file system (since Linux is case sensitive and Windows is not).

**Comment by Dr-Irv** [comment]
Look above. I moved `typing/pandas` into `pandas-stubs`.  I skipped the install step. So then it finds the local copy, finding nothing in `typings`.  Then I could change the result by renaming `py.typed` in that local copy. 

So the local copy is used.  The CI is using Linux and pip (not conda) so that takes conda out of the equation.  Could be a Python 3.9 (what I used) vs Python 3.10 but I doubt that. 

My hunch is that your methodology for combining the `pandas-stubs` and `pandas` repo works differently on Mac vs Windows and Linux. Possibly directories appear in a different order.

**Comment by erictraut** [comment]
> So then it finds the local copy

It should not find a local copy in this case. I've confirmed that it does not on my machine. A folder ending in "-stubs" is used for import resolutions only if it is installed in the current Python environment (i.e. within the `site-packages` directory). A local directory called `pandas-stubs` won't be used for resolving an import from `pandas`.

> Possibly directories appear in a different order.

That's an interesting theory. Looking at the code, I don't see an obvious way that could happen, but it would fit the symptoms. Then again, if there are platform differences, I'd expect Mac and Linux to act the same with Windows as the outlier.

**Comment by erictraut** [comment]
That's an interesting clue and gives me an idea.

Please try the following. From the command line, start the Python repl. Then type:

```
>>> import sys
>>> print(sys.path)
```

You should see a list that looks something like this:
'/Library/Frameworks/Python.framework/Versions/3.9/lib/python39.zip',
'/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9',
'/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload',
'/venvs/temp/lib/python3.9/site-packages'

Pyright determines the search paths of your Python environment by running a small script which is effectively the same as this. I would _not_ expect to see a path that points to your project directory, but given the behavior you're seeing, I wonder if your `sys.paths` includes such a path. That would explain why a local `pandas-stubs` is used to resolve an `import pandas` statement.

**Comment by Dr-Irv** [comment]
```

C:\Anaconda3\envs\base39\python39.zip
C:\Anaconda3\envs\base39\DLLs
C:\Anaconda3\envs\base39\lib
C:\Anaconda3\envs\base39
*REDACTED*\pandas-issue\venv
*REDACTED*\pandas-issue\venv\lib\site-packages
```
Note that first entry is an empty string.  Next 4 are from the base conda environment that I used to create the virtual environment (`venv`)

I think the current directory is always on the path.  Having said that, `mypy` doesn't use `pandas-stubs` that is on the path.  So this issue of trying things with `pandas-stubs` in the directory where `pyright` is being run is a bit of red herring.

Going back to installing the wheel, the files from the wheel are identical to what is in the `typings/pandas` directory.  Again, if I go to `venv/Lib/site-packages/pandas-stubs`, if I rename `py.typed` there, then things work.  If I leave it in there, then it doesn't work.  The `py.typed` file in `pandas-stubs` is needed to make `mypy` work.  Main issue seems to be that with `py.typed`, `pyright` is picking up `pandas/_libs/tslibs/timestamps.pyi` instead of `pandas-stubs/_libs/tslibs/timestamps.pyi`, whereas `mypy` is picking up the version in `pandas-stubs`

**Comment by Dr-Irv** [comment]
More data.  Under Windows, I have WSL installed, which gives me a Ubuntu environment.  After a bit of work, I was able to get python 3.9 installed, and run the script you provided.  I get the SAME error that I get on Windows.  And if I go into `/home/irv/pandas-issue/venv/lib/python3.9/site-packages/pandas-stubs` and rename `py.typed` to `py.typed.keep`, then the error goes away.

So I think you need to get access to Windows or Linux and see what is happening there, or tell me something else to look at.

Here is also a simple test to run.  Use the following file `testfile.py`:
```python
import pandas as pd

ts = pd.Timestamp("2022-06-07")
reveal_type(ts)
print(ts)
```

With the `pandas-stubs` installed, and `py.typed` there, I get the following incorrect output when typing `pyright testfile.py`:
```
/home/irv/testfile.py:4:13 - information: Type of "ts" is "Timestamp | NaTType"
```

If I rename `py.typed` as stated above, I get the right answer:
```
/home/irv/testfile.py:4:13 - information: Type of "ts" is "Timestamp"
```

**Comment by Dr-Irv** [comment]
One more piece of data.  It seems that if `py.typed` contains the string `partial` and is in the `pandas-stubs` folder, that causes the problem.  If I just have an empty `py.typed` file in `pandas-stubs`, then `pyright` and `mypy` both work.

**Comment by jakebailey** [comment]
> I think the current directory is always on the path. 

IIRC we ignore this in the editor entirely, filtering out the value and instead using the workspace root as the "current" directory +/- the fancy parent-folder-resolution heuristic), so I don't know that this is the problem.

I haven't been following along too much, but I did try the first set of instructions on https://github.com/microsoft/pyright/issues/3540#issuecomment-1147079578 on my Windows machine, and I get type errors (some mention `NaTType`), so I think I am reproducing the failure?

Doing the simple test case:

```
npx pyright -p .\pyrightdistconfig.json .\tests\testfile.py
npx: installed 1 in 4.451s
Loading configuration file at E:\work\python\pyright3540\pandas-issue\pandas-stubs\pyrightdistconfig.json
Assuming Python version 3.10
Assuming Python platform Windows
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath E:\work\python\pyright3540\pandas-issue\pandas-stubs\typings is not a valid directory.
Searching for source files
Found 1 source file
E:\work\python\pyright3540\pandas-issue\pandas-stubs\tests\testfile.py
  E:\work\python\pyright3540\pandas-issue\pandas-stubs\tests\testfile.py:4:13 - information: Type of "ts" is "Timestamp | NaTType"
0 errors, 0 warnings, 1 information
Completed in 0.818sec
```

I can dump whatever super-verbose thing you'd like, if that helps.

**Comment by jakebailey** [comment]
Alright, so I performed those steps I did before on Windows, Linux, and macOS, and it's the case that Windows and Linux fail, while macOS passes. I'm going to see if I can do a little debugging / verbose log diffing to see if there's something obvious here.

**Comment by erictraut** [comment]
Ah, that explains it. Thanks Jake! I should be able to move forward with my diagnosis then by recreating the situation that exists on Windows and Linux.

@Dr-Irv, do you happen to know why these stub files are not installed on MacOS? Is it just an oversight, or was that intentional? Just curious.

**Comment by Dr-Irv** [comment]
I just looked on a colleague's Mac, and the PYI files are there in `site-packages/pandas/_libs/tslibs/` and `site-packages/pandas/_libs/`   I asked him to list all the PYI files there and compare them to the PYI files in my Windows installation, and they are all the same.

I think the issue is that on the Mac, you are resolving `pandas._libs.lib` to the SO File, and not the PYI file.  On Windows, there is no SO file.  There are `pyd` files.

**Comment by jakebailey** [comment]
The pyi files are definitely not in my install:

```
(venv) jabaile@Jakes-MacBook-Pro pandas-stubs % ls ../venv/lib/python3.9/site-packages/pandas/_libs 
__init__.py				missing.cpython-39-darwin.so
__pycache__				ops.cpython-39-darwin.so
algos.cpython-39-darwin.so		ops_dispatch.cpython-39-darwin.so
arrays.cpython-39-darwin.so		parsers.cpython-39-darwin.so
groupby.cpython-39-darwin.so		properties.cpython-39-darwin.so
hashing.cpython-39-darwin.so		reduction.cpython-39-darwin.so
hashtable.cpython-39-darwin.so		reshape.cpython-39-darwin.so
index.cpython-39-darwin.so		sparse.cpython-39-darwin.so
indexing.cpython-39-darwin.so		testing.cpython-39-darwin.so
internals.cpython-39-darwin.so		tslib.cpython-39-darwin.so
interval.cpython-39-darwin.so		tslibs
join.cpython-39-darwin.so		window
json.cpython-39-darwin.so		writers.cpython-39-darwin.so
lib.cpython-39-darwin.so
```

**Comment by Dr-Irv** [comment]
> The pyi files are definitely not in my install:

My colleague used `conda` to do the install, and I guess you are using `pip`.  So I looked in the wheel file and the conda file for MacOS for pandas 1.4.2, and it seems that the `.pyi` files are not in the pip wheel for MacOS.  I've opened an issue for that.  https://github.com/pandas-dev/pandas/issues/47288

**Comment by Dr-Irv** [comment]
So if you want to replicate this on a Mac, download https://anaconda.org/conda-forge/pandas/1.4.2/download/osx-64/pandas-1.4.2-py39he845741_2.tar.bz2 and then do
```bash
bunzip pandas-1.4.2-py39he845741_2.tar.bz2
cp pandas/_libs/*.pyi  <site-packages>/python3.10/pandas/_libs
cp pandas/_libs/tslibs/*.pyi  <site-packages>/python3.10/pandas/_libs/tslibs
```
where `<site-packages>` is your location of your library files.  I then think you would replicate on the Mac what @jakebailey and I were seeing on Windows and Linux.

**Comment by Dr-Irv** [comment]
Another way to replicate this on a Mac is to pick up a MacOS wheel for pandas 1.4.1 from pypi.org . Those have the PYI files in them.  It seems something went wrong in the 1.4.2 process that we most likely won't fix, since 1.4.3 will be released next week

**Comment by erictraut** [comment]
Thanks @jakebailey and @Dr-Irv, I'm able to repro the problem with pandas 1.4.1. Investigating further.

**Comment by erictraut** [comment]
Here's a PR that addresses this issue: https://github.com/microsoft/pyright/pull/3568.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.254, which I just published. It will also be included in the next release of pylance.
